doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,706 @@
+
+
+Backward compatibility
+----------------------
+
+While the actual backward compatibility impact is not yet fully clear,
+we anticipate, and accept the following changes:
+
+* **Python API**:
+
+  * ``type(np.dtype("f8"))`` will be a subclass of ``np.dtype``, while right
+    now ``type(np.dtype("f8")) is np.dtype``.
+    Code should use ``isinstance`` checks, and in very rare cases may have to
+    be adapted to use it.
+
+* **C-API**:
+
+    * In old versions of NumPy ``PyArray_DescrCheck`` is a macro which uses
+      ``type(dtype) is np.dtype``. When compiling against an old NumPy version,
+      the macro may have to be replaced with the corresponding
+      ``PyObject_IsInstance`` call. (If this is a problem, we could backport
+      fixing the macro)
+
+   * The UFunc machinery changes will break *limited* parts of the current
+     implementation. Replacing e.g. the default ``TypeResolver`` is expected
+     to remain supported for a time, although optimized masked inner loop iteration
+     (which is not even used *within* NumPy) is expected to not remain supported
+     and lead to errors instead.
+
+   * All current functions defined on the dtypes currently, such as
+     ``PyArray_Descr->f->nonzero``, will move to a new home and recieve new
+     ways to access them.
+     This means that in the long run some (rare) lowlevel access code will
+     have to be changed to use the new API.
+
+* **dtype implementors (C-API)**:
+
+  * The array that is currently provided to some functions (such as cast functions),
+    may not be provided anymore generally.
+    For example ``PyArray_Descr->f->nonzero`` or ``PyArray_Descr->f->copyswapn``,
+    may instead receive a dummy array object with only some fields (mainly the
+    dtype), being valid.
+    At least in some code paths, a similar mechanism is already used.
+
+  * The ``scalarkind`` slot and registration of scalar casting will be
+     removed/ignored without replacement.
+     It currently allows partial value based casting.
+     The ``PyArray_ScalarKind`` function will continue to work for builtin types,
+     but will not be used internally and be deprecated.
+
+   * Currently user dtypes are defined as instances of ``np.dtype``.
+     The creation works by the user providing a prototype instance.
+     NumPy will need to modify at least the type during registration.
+     This has no effect for either ``rational`` or ``quaternion`` and mutation
+     of the structure seems unlikely after registration.
+
+Since there is a fairly large API surface concerning datatypes, further changes
+or the limitation of a certain function to currently existing datatypes is
+likely to occur.
+For example functions which use the type number as input
+should be replaced with functions taking DType classes instead.
+Although public, large parts of this C-API seems very rarely and possibly
+never used by downstream projects.
+
+
+
+
+Detailed Description
+--------------------
```

Comment:
It took me a couple readings to figure out that the "Detailed Description" subheadings correlated to the "design decisions" made in the "Scope" section. Consider explicitly stating so here in the introduction to the section, or at least linking the subheadings below back to the original listing of the design decisions.